Adding recovery services backup swagger json

dragonfly91 opened the pull request:
Adding recovery services backup swagger json

Commit:
Adding recovery services backup swagger json

Adding recovery services backup swagger json

azurecla commented:
Hi __@dragonfly91__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (anudeeb). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+        ],
+        "description": "Provides the details of the backed up item. This is an asynchronous operation. To know the status of the operation, call the GetItemOperationResult API.",
+        "operationId": "ProtectedItems_GetProtectedItem",
+        "consumes": [],
```
Remove empty properties

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "ProtectedItems"
+        ],
+        "description": "Provides the details of the backed up item. This is an asynchronous operation. To know the status of the operation, call the GetItemOperationResult API.",
+        "operationId": "ProtectedItems_GetProtectedItem",
```
In AutoRest, we split the operation id on the underscore character and expect that it will follow the pattern "Noun_Verb". We generally recommend avoiding duplicating the noun in the verb part too: https://github.com/Azure/autorest/wiki/operation-id-noun-in-verb

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "ProtectedItems"
+        ],
+        "description": "Lists all items that can be backed up within a subscription.",
+        "operationId": "ProtectedItems_ListProtectedItems",
```
(same goes for all the operation ids in this file)

tbombach replied:
In our [best practices document](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md), we encourage consistency in the names of `List` methods across services. For listing by resource group name, we have the pattern `[Noun]_ListByResourceGroup`.

Is there a way to list all `ProtectedItems` or `Jobs` for a subscription (across all resource groups)?. I don't see an operation for that in this spec. The best practice is to name that operation `[Noun]_List`

dragonfly91 replied:
@tbombach This operation actually retrieves the list of Items that are protected by a vault. Protected Items always exist in the context of a vault, and thus a particular resource group - specifically, the rg where the vault resides. However, this is not the intention of the API here - this API lists all the items protected by the vault. So, can we keep the operation ID as ProtectedItems_List?

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+            "$ref": "#/parameters/ApiVersion"
+          },
+          {
+            "name": "resourceName",
```
For common parameter definitions like this, it's recommended to put it in the global parameters section (like you've done with `SubscriptionId`). Since AutoRest will create a property on the service client by default, you can customize this behavior with the [`x-ms-parameter-location` extension](https://github.com/Azure/autorest/blob/master/Documentation/swagger-extensions.md#x-ms-parameter-location) to make it remain a method parameter.

I realize this isn't in the [extensions documentation](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-extensions.md) for this repo, so I'll work to make sure it's put there. In general, we're looking to improve our documentation to make it easier to discover best practices like this. We'd like to make it more "scenario focused" and discoverable for service teams writing Swagger. Right now, you kinda need to know about all of the extensions to know which ones should be applied. Any feedback about our documentation is more than welcome!

dragonfly91 replied:
@tbombach I understand your concern here. But, the problem with making resourceGroupName and resourceName global is that - the client would be tied to a particular vault and all the operations will be done only in the context of that global vault / resource. This restricted the way we could run our tests - where we wanted to create multiple vaults via the same client. Is it possible to keep these as it is?

amarzavery replied:
I am not able to understand the comment.
What we are saying is:
> you can put common parameters in the global parameters section and add the extension "x-ms-parameter-location": "method" for parameters like resourcegroupname and resourcename.  By default Autorest will make all the global parameters as properties on the client. Hence you need to add that extension

dragonfly91 replied:
Yes, I understand the idea of making them global parameters - then they will become properties of the client. But the issue is that, we need to reset the properties in the tests, before calls, in order to run a particular operation in a different vault's context. My question is - can this be avoided? Hope I'm clear :)

amarzavery replied:
No, you are still not getting it. 
We are saying just define resourceGroupName and resourceName once in the global parameters section and apply the `"x-ms-parameter-location": "method"` extension on them. In that way they will not be properties on the client.

By defining them once, you can reference them at multiple places in the spec there by avoiding duplication.

dragonfly91 replied:
Oh okay, I got it now. Will implement it, thanks!

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+            "$ref": "#/parameters/ResourceNamespace"
+          },
+          {
+            "name": "resourceGroupName",
```
Same goes for the `"resourcesGroupName"` parameter definition

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      }
+    },
+    "/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceNamespace}/vaults/{resourceName}/backupProtectedItems": {
+      "get": {
```
Please add the [`x-ms-pageable` extension](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-extensions.md#x-ms-pageable) for this operation if it's pageable. This will direct AutoRest to generate `Next()` methods that make it easier to get the next page of data and the return type for the method will be an `IPageable` collection of items.

dragonfly91 replied:
This is not a pageable operation. We've added this extension wherever it was applicable.

amarzavery replied:
@dragonfly91 - As per the [definition of the model](https://github.com/Azure/azure-rest-api-specs/pull/495/files/4cce22c313c64a572bc9d05b535e63a15e6461ad#diff-06bfd98e833d5fac8ee1c2cc16c1621cR642) in 200 response, it has a nextLink property. Hence it very much looks like a pageable operation.

On another note, if you have a list operation that is not pageable but returns an object that has a value property of type array then you should still apply the x-ms-pageable extension with nextLink explicitly set to null. This will indicate Autorest to return an IEnumerable<T> where T is the type of item returned in the value array. 

All of this is very explained in the x-ms-pageable extension. Please read it over [here](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-extensions.md#x-ms-pageable).

dragonfly91 replied:
@amarzavery You are correct, this is indeed a pageable operation. Will add the x-ms-pageable extension. However, we didn't find a case where we might have to set this value to null, yet!

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      }
+    },
+    "/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceNamespace}/vaults/{resourceName}/backupJobs": {
+      "get": {
```
Please use `x-ms-pageable` if this operation is pageable

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+        "tags": [
+          "Job"
+        ],
+        "description": "Provides list of jobs.",
```
If this operation is pageable, it would be helpful to put that in the description (e.g. "Provides a pageable list of jobs.")

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+        }
+      }
+    },
+    "Resource[Job]": {
```
Rather than putting the type information in the name (i.e. `Resource[Job]`), I would recommend changing the definitions structure a bit:

You could have a base `Resource` type that has all of the Resource properties (e.g. `id`, `name`). Please include the [`x-ms-azure-resource` extension] to make sure that this inherits from the `Resource` type from the client runtime. It helps customers, since `Resource` is common across all services.

`ProtectedItem` and `Job` definitions can then inherit from the `Resource` definition using JSON allOf: https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/creating-swagger.md#model-inheritance. That way `ProtectedItem` and `Job` have the Resource properties and don't re-define them.

amarzavery replied:
This should be named as "Job".
All the generic resource properties should be modeled in the "Resource" model definition with location being required as per the ARM spec. Take a look at the arm-storage spec from this repo as an example

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+    "application/octet-stream"
+  ],
+  "paths": {
+    "/Subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{resourceNamespace}/vaults/{resourceName}/backupFabrics/{fabricName}/protectionContainers/{containerName}": {
```
resourceNamespace shouldn't be a path parameter. It used to be a parameter in the old specs. However, in this structure all path parameters are required. This should simply be replaced with the string "RecoveryServices"

dragonfly91 replied:
Hi @amarzavery , in our internal test framework, we have multiple test deployments identified by the resource namespace. For example RecoveryServicesBVTD etc. This will become a major pain point for us if we hard-code this value here in all the paths. Is it possible to treat this as an exception? If not, can you suggest a way where we can set the resource namespace during run-time?

amarzavery replied:
Sure I very well understand your concern. However, one cannot mess up the spec, for ease of use in testing internally. We target for customer's ease of use. There are different ways in which you can solve this problem.
One thing that can be done is to add a delegating handler to the request pipeline that changes the resourceNamespace to the desired value, before the request is sent. In node.js you would add a filter to the request pipeline to do the same.

Another thing that could be done is to create a manual wrapper on top of the generated code that can be used for internal testing.

dragonfly91 replied:
That sounds promising. Will explore that route, thanks!

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/Resource[ProtectionContainer]"
```
Please make a change in the swashbuckle plugin to name Model Definitions correctly. This should actually be named as "ProtectionContainer"

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+            }
+          }
+        },
+        "deprecated": false
```
Can we get away with this useless property.

tbombach replied:
This is the default value, no need to include it.

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+        "tags": [
+          "ProtectedItems"
+        ],
+        "description": "Provides the details of the backed up item. This is an asynchronous operation. To know the status of the operation, call the GetItemOperationResult API.",
```
To know the status of the operation, call the GetItemOperationResult API." - This statement in the description does not make sense any more. When you mark the operation with the long running operation the AutoRest generated code will poll the operation until it reaches one of the terminal states (success, failure, cancelled). This is a different approach than the old code generator.

tbombach replied:
Is this call long-running? It seems strange for the "get" call to be an async operation that you have to call another API to get the status of the "get"

dragonfly91 replied:
Yes, but we want to let the clients do the polling as of now. Is that fine?

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+            "$ref": "#/parameters/ApiVersion"
+          },
+          {
+            "name": "resourceName",
```
instead of naming this resourceName, it would be nice if this is named as vaultName. That makes things more clear.

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "200": {
+            "description": "OK",
+            "schema": {
+              "$ref": "#/definitions/Resource[ProtectedItem]"
```
ProtectedItem

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "Job"
+        ],
+        "description": "Exports all jobs for a given SAS URL. The SAS URL will expire within 15 minutes of its creation.",
+        "operationId": "Job_ExportJobs",
```
We recommend the operation group name (part before the underscore) to be plural.

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "BackupOperationResult"
+        ],
+        "description": "Fetches the status of an operation such as triggering a backup, restore. The status can be in progress, completed or failed. You can refer to the OperationStatus enum for all the possible states of an operation. Some operations create jobs. This method returns the list of jobs when the operation is complete.",
+        "operationId": "BackupOperationResult_GetBackupOperationStatus",
```
Is this API, exposed for polling long running operations. If yes, then this wont be required as AutoRest generated code will do the polling for the customer. However, if you still want to expose this then that is perfectly fine as well.

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          {
+            "name": "fabricName",
+            "in": "path",
+            "description": "Name of the fabric where the container belongs",
```
Please be consistent with periods at the end of descriptions.

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "description": "Client Api Version.",
+      "required": true,
+      "type": "string",
+      "default": "RecoveryServices"
```
incorrect default value

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+    "SubscriptionId": {
+      "name": "subscriptionId",
+      "in": "path",
+      "description": "ID of the subscription in which the recovery services vault is present.",
```
"The subscription Id." - is simple and clear

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "BackupOperationResult"
+        ],
+        "description": "Provides the status of the delete operations such as deleting backed up item. Once the operation has started, the status code in the response would be Accepted. It will continue to be in this state till it reaches completion. On successful completion, the status code will be OK. This method expects OperationID as an argument. OperationID is part of the Location header of t ...
+        "operationId": "BackupOperationResult_GetBackupOperationResult",
```
same as above

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+    }
+  },
+  "definitions": {
+    "Object": {
```
Why do you need this?

amarzavery replied:
Wouldn't this conflict in languages like C#, Java and javascript where all objects are derived from Object?

tbombach replied:
Unused definition

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+        },
+        "status": {
+          "description": "Operation status. Possible values: InProgress, Succeeded, Failed, Canceled.",
+          "type": "string"
```
make this an enum and apply the x-ms-enum enxtension. You can find more about this extension from the documentation folder of this repo.

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+        },
+        "startTime": {
+          "format": "date-time",
+          "description": "Operation start time.",
```
Please specify in the description the format of the value; as in ISO-8601. This makes things more clear for the customer

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "properties": {}
+    },
+    "ResourceList[Job]": {
+      "description": "Resource list wrapper for \"T\" type",
```
Please don't use swashbuckle as the generated spec oozes with Csharpiness all over. Provide a better description

amarzavery replied:
The name of the model definition should be JobListResult

dragonfly91 replied:
Hi @amarzavery , we've adopted the swashbuckle approach as we hope that it'll make the job of creating the swagger spec seamless - by automation. If the description is the issue here, it can be fixed. Any other reasons why swashbuckle may be avoided?

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          }
+        ],
+        "responses": {
+          "202": {
```
Is this a [long running operation](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/creating-swagger.md#enable-asynchronous-operations-with-x-ms-long-running-operation)? If so, you can mark it with `x-ms-long-running-operation` and define a terminal response code state (i.e. `200`) in the responses section with the type that eventually returns.

dragonfly91 replied:
Yes, but we want to let the clients do the polling as of now. Is that fine?

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "type": "object",
+      "properties": {
+        "value": {
+          "description": "Gets or sets the list of resources.",
```
Please remove "gets or sets" from everywhere. AutoRest will add this appropriately as per it's immutability,

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "ProtectionContainers"
+        ],
+        "description": "Gets details of the specific container registered to your Recovery Services Vault.",
+        "operationId": "ProtectionContainers_BMSGetContainer",
```
BMS? Could this operation id be "ProtectionContainers_Get"?

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "type": "string"
+        },
+        "properties": {
+          "$ref": "#/definitions/Job",
```
Please apply the `"x-ms-client-flatten": true` extension over here. This makes things easy for the customer as he/she does not have to access `Job.properties.foo` every time.

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "type": "string"
+        },
+        "backupManagementType": {
+          "description": "Backup management type to execute the current job. Gets or sets the BackupManagementType to IaasVM, DPM or MAB",
```
This should be an enum for better discoverability

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "type": "object",
+      "properties": {
+        "status": {
+          "description": "Status of the job. \r\n            Possible values: InProgress, Completed, Failed, CompletedWithWarnings, Cancelling, Cancelled.",
```
Please remove \r and \n from the description. AutoRest will automatically format it.

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "properties": {
+        "status": {
+          "description": "Status of the job. \r\n            Possible values: InProgress, Completed, Failed, CompletedWithWarnings, Cancelling, Cancelled.",
+          "type": "string"
```
this should be an enum.

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      }
+    },
+    "ResourceList[ProtectedItem]": {
+      "description": "Resource list wrapper for \"T\" type",
```
ProtectedItemListResult and better description

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "description": "Resource wrapper for \"T\" type",
+      "type": "object",
+      "properties": {
+        "id": {
```
same comments as above

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+        },
+        "itemType": {
+          "description": "Type of workload this item represents. \r\n            Possible values: VM, FileFolder.",
+          "type": "string"
```
enum?

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "properties": {
+        "backupManagementType": {
+          "description": "Backup management type for the backed up item. Possible values: AzureIaasVM, MAB, DPM.",
+          "type": "string"
```
enum?

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+            "$ref": "#/parameters/SubscriptionId"
+          },
+          {
+            "name": "$filter",
```
Is OData filter allowed on this post operation? I imagine it's possible, if it filters which jobs can be exported, but I just wanted to verify.

dragonfly91 replied:
Yes, it is possible.

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "properties": {
+        "expand": {
+          "description": "Specifies if the additional information should be provided for this item.",
+          "type": "string"
```
Are all the properties expandable? How would the user know what properties can one expand on?  Please mention this in the description to make customer's life easy.

dragonfly91 replied:
@amarzavery The expand query parameter here just takes True or False. When expand is True, the object returned would contain some extended properties.

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "description": "Resource wrapper for \"T\" type",
+      "type": "object",
+      "properties": {
+        "id": {
```
same comment as above

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+        },
+        "backupManagementType": {
+          "description": "Type of backup managemenent for the container. Possible values: AzureIaasVM, DPM or MAB",
+          "type": "string"
```
enum

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "type": "string"
+        },
+        "backupItemType": {
+          "description": "Backup items associated with this container.",
```
description seems incorrect. Property name suggests that it is some type. If so then should this be an enum?

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+        },
+        "protectionState": {
+          "description": "Backup state of the backed up item. Possible values: Protected, ProtectionStopped, IRPending or ProtectionError",
+          "type": "string"
```
enum

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "properties": {
+        "oldestRecoveryPoint": {
+          "format": "date-time",
+          "description": "The oldest backup copy available for this item in the service.",
```
Please mention the format of the date-time value everywhere- a general comment

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      "properties": {
+        "duration": {
+          "description": "Time elapsed during the execution of this job.",
+          "type": "string"
```
"format": "duration"

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+        "tags": [
+          "BackupOperationResult"
+        ],
+        "description": "Provides the status of the delete operations such as deleting backed up item. Once the operation has started, the status code in the response would be Accepted. It will continue to be in this state till it reaches completion. On successful completion, the status code will be OK. This method expects OperationID as an argument. OperationID is part of the Location header of t ...
```
I didn't seem to find an operation that the user would use to start the delete operation mentioned in this description. I would expect that operation to also be marked as `x-ms-long-running-operation` if it followed the guidelines for long-running operation polling: https://github.com/Microsoft/api-guidelines/blob/master/Guidelines.md#1327-the-typical-flow-polling

dragonfly91 replied:
We have a requirement in our clients to have the clients handle the polling on their own. Hence, we explore the route of making this long-running. Is this not allowed with swagger?

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "type": "string"
+        },
+        "workloadType": {
+          "description": "Workload type of backup item",
```
What are the possible values?

amarzavery commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+          "type": "string"
+        },
+        "mabServerType": {
+          "description": "Server type of MAB container",
```
possible values

tbombach commented on arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:
```diff
@@ -0,0 +1,1548 @@
+      }
+    },
+    "OperationStatusExtendedInfo": {
+      "description": "Base class for additional information of operation status.",
```
We shouldn't have base classes without properties. If it doesn't provide any properties for its children to inherit, it shouldn't be included.

Commit:
Review comments and complete set of operations

Addressed the review comments in the previous iteration.
Added some missing operations.

dragonfly91 commented:
@amarzavery @tbombach I've tried to address most of the comments given by you in the first iteration. A few comments left:
1. Remove empty Object - swashbuckle tool limitation.
2. Remove deprecated : false property - swashbuckle tool limitation.
3. Should not have empty base classes - will take it up a little later, as we have to make some code changes.

For the points 1 and 2, can these be overlooked? Otherwise, we need to find a way to manually edit the generated swagger JSON doc. 

In the meanwhile, can you continue with the review?

amarzavery commented:
>         ],

please mark this as a long running operation using the extension.

---
Refers to: arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:35 in 703b99e. [](commit_id = 703b99efce129886469fc370fa47a373f958b4ae, deletion_comment = False)

amarzavery commented:
>         "responses": {

I think a 200 response should also be defined and it should have a schema.

---
Refers to: arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:78 in 703b99e. [](commit_id = 703b99efce129886469fc370fa47a373f958b4ae, deletion_comment = False)

amarzavery commented:
>           "RecoveryPoint"

Please make sure the tags match the Noun of the OperationId.

---
Refers to: arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:28 in 703b99e. [](commit_id = 703b99efce129886469fc370fa47a373f958b4ae, deletion_comment = False)

amarzavery commented:
>         "description": "Provisions a script which invokes an iSCSI connection to the backup data. Executing this script opens a file explorer displaying all the recoverable files and folders. This is an asynchronous operation. To know the status of provisioning, call GetProtectedItemOperationResult API.",

> "Executing this script opens a file explorer displaying all the recoverable files and folders."

I am not able to understand the above statement in the context of this operation. This is a REST API call and not a script. Are you saying that this effect (n file explorers will be opened) will be on the container/machine after n REST API calls are made? This sounds very funny to me.

---
Refers to: arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:91 in 703b99e. [](commit_id = 703b99efce129886469fc370fa47a373f958b4ae, deletion_comment = False)

dragonfly91 commented:
This API would download a script, which when run, will invoke an instance of the file explorer.

From: Amar Zavery [mailto:notifications@github.com]
Sent: Saturday, September 10, 2016 7:25 AM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Anudeep Bangari <anudeeb@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] Adding recovery services backup swagger json (#495)


    "description": "Provisions a script which invokes an iSCSI connection to the backup data. Executing this script opens a file explorer displaying all the recoverable files and folders. This is an asynchronous operation. To know the status of provisioning, call GetProtectedItemOperationResult API.",

"Executing this script opens a file explorer displaying all the recoverable files and folders."

I am not able to understand the above statement in the context of this operation. This is a REST API call and not a script. Are you saying that this effect (n file explorers will be opened) will be on the container/machine after n REST API calls are made? This sounds very funny to me.

________________________________

Refers to: arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:91 in 703b99e<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fAzure%2fazure-rest-api-specs%2fcommit%2f703b99efce129886469fc370fa47a373f958b4ae&data=02%7c01%7canudeeb%40microsoft.com%7c69c71b45af1a40a6616208d3d91d6abd%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636090692769162260&sdata=Eo5HYDpNZnYeqqTwpOiWF0rvRWFHHMPYwdPiDovBIeE%3d>.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fAzure%2fazure-rest-api-specs%2fpull%2f495%23issuecomment-246081032&data=02%7c01%7canudeeb%40microsoft.com%7c69c71b45af1a40a6616208d3d91d6abd%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636090692769162260&sdata=ZW3iBG9oEPuaUSxUZ%2bH4q%2f2E0ZbCxYQcRVcKZS0J8lg%3d>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fnotifications%2funsubscribe-auth%2fABHpO_uf9bh0UhOeagR6UOXZpFR50d4tks5qog3agaJpZM4JuJYj&data=02%7c01%7canudeeb%40microsoft.com%7c69c71b45af1a40a6616208d3d91d6abd%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636090692769162260&sdata=y89kQd88ArBHc26KTgEOc4Vqte8cdsro1i8TwRRk3vI%3d>.

dragonfly91 commented:
This API (and other similar ones) are also being used by our Ibiza portal extension (Recovery Services) and we need the capability to poll the long running operation ourselves and not depend on the AutoRest’s polling feature. So, we can’t mark this operation as long running using the extension.

From: Amar Zavery [mailto:notifications@github.com]
Sent: Saturday, September 10, 2016 6:11 AM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Anudeep Bangari <anudeeb@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] Adding recovery services backup swagger json (#495)


    ],

please mark this as a long running operation using the extension.

________________________________

Refers to: arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:35 in 703b99e<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fAzure%2fazure-rest-api-specs%2fcommit%2f703b99efce129886469fc370fa47a373f958b4ae&data=02%7c01%7canudeeb%40microsoft.com%7c0de451d6a75141d1d6b908d3d9132bac%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636090648766204067&sdata=Yynowd58jScAPIFZZhGwjTMidGdT6qdOxnjmx2zaD7c%3d>.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fAzure%2fazure-rest-api-specs%2fpull%2f495%23issuecomment-246075679&data=02%7c01%7canudeeb%40microsoft.com%7c0de451d6a75141d1d6b908d3d9132bac%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636090648766204067&sdata=bA%2bnwcm65QROcGQdh64ZIn3HyAzJaxrua1V%2f9fuQsSk%3d>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fnotifications%2funsubscribe-auth%2fABHpO9QErHhqoAFOli7Wzxr5DHVOMsuEks5qofyqgaJpZM4JuJYj&data=02%7c01%7canudeeb%40microsoft.com%7c0de451d6a75141d1d6b908d3d9132bac%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636090648766204067&sdata=yCxaXncl9P9GcyEt2PAOpkrtyao2MfqClcr%2f4tYWZys%3d>.

dragonfly91 commented:
This is a long running operation on the service side, and thus it won’t return a 200. This was exactly how this particular API was modeled in hydra as well.

From: Amar Zavery [mailto:notifications@github.com]
Sent: Saturday, September 10, 2016 6:14 AM
To: Azure/azure-rest-api-specs <azure-rest-api-specs@noreply.github.com>
Cc: Anudeep Bangari <anudeeb@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-rest-api-specs] Adding recovery services backup swagger json (#495)


    "responses": {

I think a 200 response should also be defined and it should have a schema.

________________________________

Refers to: arm-recoveryservicesbackup/2016-06-01/swagger/recoveryservicesbackup.json:78 in 703b99e<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fAzure%2fazure-rest-api-specs%2fcommit%2f703b99efce129886469fc370fa47a373f958b4ae&data=02%7c01%7canudeeb%40microsoft.com%7c7002b11509e94e9e19ff08d3d9138b02%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636090650361756459&sdata=11WbXtvr%2bzLywzvzI%2b4IWAX13AtCrMwPgtqJ8vNGAoA%3d>.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fAzure%2fazure-rest-api-specs%2fpull%2f495%23issuecomment-246075888&data=02%7c01%7canudeeb%40microsoft.com%7c7002b11509e94e9e19ff08d3d9138b02%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636090650361756459&sdata=tgjOVCF5YDMHOSJB87CfsLXH8Mz15Mpp6piQTo%2fHs7Y%3d>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3a%2f%2fgithub.com%2fnotifications%2funsubscribe-auth%2fABHpO5CIXpBC3Ig3ZmywSM-G0zrTsdcsks5qof1IgaJpZM4JuJYj&data=02%7c01%7canudeeb%40microsoft.com%7c7002b11509e94e9e19ff08d3d9138b02%7c72f988bf86f141af91ab2d7cd011db47%7c1%7c0%7c636090650361756459&sdata=qD8WPbw3o1uqBBeqwz64SKepC%2bACUCLAKVJRIfYC3D4%3d>.

Commit:
Implemented review comments

1. Fixed the empty base class issue - this was due to a bug in the
swagger doc generation process in one of Swashbuckle's filters.
2. Modified Tags to match the Noun in OperationId

dragonfly91 commented:
Hi, in the latest commit, we've addressed all the comments that had an action item in it. Please review / merge the PR!

tbombach commented:
Merged

AutorestCI commented:
No modification for Python

AutorestCI commented:
No modification for NodeJS

AutorestCI commented:
No modification for Ruby
